### PR TITLE
feat(cna): enhance CLI interaction

### PR DESCRIPTION
**Pull request by devjiwonchoi**
cc @samcx Inspired by your work #61877, enhanced usage of `commander` for CNA as well.

**NOTE: This partial change focuses on the main `index.ts` file. The following PR will enhance the rest.**

## Key Changes

- Upgraded `commander` from `v2.20.0` to `v12.0.0`
- Added support for an internal flag `--dry-run` that returns data without creating an app for debugging and testing
- Arrange prompt order by importance: app > ts > eslint > tailwind > src > import alias

### Visual

> Added console logger used in `next` for consistency

#### After

Screenshot 2024-04-05 at 6 24 17 PM

#### Before

Screenshot 2024-04-05 at 6 29 30 PM

<details><summary>Help Message</summary>

### After

Screenshot 2024-03-29 at 2 14 00 AM

### Before

Screenshot 2024-03-29 at 1 56 08 AM

</details> 

### Interaction

<details><summary>Ask for confirmation before resetting the saved preferences</summary>

### After

https://github.com/vercel/next.js/assets/120007119/b5a2546f-c8b7-47bc-a7c1-27fc5eacaa54

### Before

https://github.com/vercel/next.js/assets/120007119/9792c81d-34d2-4217-85a4-2542a6a842d1

</details> 

<details><summary>Validate import alias pattern</summary>

### After

https://github.com/vercel/next.js/assets/120007119/6a9b3e5a-655c-4fff-83d5-e5ffce1d2705

### Before

https://github.com/vercel/next.js/assets/120007119/dbd52d4e-5e49-4242-b7f5-0165949a8366

</details> 

### Aliases

- Added `--reset` for `--reset-preferences`
- Added `--use` for `--use-<package-manager>`
  - Since `commander` has a handy option `choices`, we don't need to handle the package managers explicitly.

## RFC

I plan to move (should support legacy also) example flags (`-e, --example`, `--example-path`) as a separate command and enhance handling:

```sh
create-next-app example <github-url | example-name>
```

We can handle it without `--example-path` and keep the flags simple.

### CODE REVIEW

**Comment by ijjk** [comment]
Allow CI Workflow Run

- [ ] approve CI run for commit: 0beb515a07517c69519e9397605790001c680291

Note: this should only be enabled once the PR is ready to go and can only be enabled by a maintainer

File: packages/create-next-app/helpers/get-pkg-manager.ts [hunk 1]
```diff
@@ -1,19 +1,10 @@
...
-  if (userAgent.startsWith('pnpm')) {
-    return 'pnpm'
-  }
-
-  if (userAgent.startsWith('bun')) {
-    return 'bun'
-  }
-
+  const userAgent = process.env.npm_config_user_agent
+  if (!userAgent) return 'npm'
```
Comment (styfle) [inline packages/create-next-app/helpers/get-pkg-manager.ts:5]: Lets keep the `process.env.npm_config_user_agent || ''` so we dont need two different `return npm`.

**Comment by ijjk** [comment]
# Stats from current PR

<details open>
<summary><strong>Default Build</strong></summary>

<br/>

<details>
<summary><strong>General</strong></summary>

|  | vercel/next.js canary  | devjiwonchoi/next.js refactor/cna | Change |
| - | - | - | - |
| buildDuration | 17.3s | 17.3s | N/A |
| buildDurationCached | 10.8s | 8.1s | N/A |
| nodeModulesSize | 199 MB | 199 MB | ✓ |
| nextStartRea..uration (ms) | 469ms | 464ms | N/A |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | devjiwonchoi/next.js refactor/cna | Change |
| - | - | - | - |
| 2453-HASH.js gzip | 31.4 kB | 31.3 kB | N/A |
| 3304.HASH.js gzip | 181 B | 181 B | ✓ |
| 3f784ff6-HASH.js gzip | 53.7 kB | 53.7 kB | ✓ |
| 8299-HASH.js gzip | 5.04 kB | 5.04 kB | N/A |
| framework-HASH.js gzip | 45.2 kB | 45.2 kB | ✓ |
| main-app-HASH.js gzip | 242 B | 242 B | ✓ |
| main-HASH.js gzip | 32.2 kB | 32.2 kB | N/A |
| webpack-HASH.js gzip | 1.68 kB | 1.68 kB | N/A |
| Overall change | 99.3 kB | 99.3 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | devjiwonchoi/next.js refactor/cna | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | devjiwonchoi/next.js refactor/cna | Change |
| - | - | - | - |
| _app-HASH.js gzip | 196 B | 197 B | N/A |
| _error-HASH.js gzip | 184 B | 184 B | ✓ |
| amp-HASH.js gzip | 505 B | 505 B | ✓ |
| css-HASH.js gzip | 324 B | 325 B | N/A |
| dynamic-HASH.js gzip | 2.5 kB | 2.5 kB | N/A |
| edge-ssr-HASH.js gzip | 258 B | 258 B | ✓ |
| head-HASH.js gzip | 352 B | 352 B | ✓ |
| hooks-HASH.js gzip | 370 B | 371 B | N/A |
| image-HASH.js gzip | 4.21 kB | 4.21 kB | ✓ |
| index-HASH.js gzip | 259 B | 259 B | ✓ |
| link-HASH.js gzip | 2.67 kB | 2.67 kB | N/A |
| routerDirect..HASH.js gzip | 314 B | 312 B | N/A |
| script-HASH.js gzip | 386 B | 386 B | ✓ |
| withRouter-HASH.js gzip | 309 B | 309 B | ✓ |
| 1afbb74e6ecf..834.css gzip | 106 B | 106 B | ✓ |
| Overall change | 6.57 kB | 6.57 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | devjiwonchoi/next.js refactor/cna | Change |
| - | - | - | - |
| _buildManifest.js gzip | 481 B | 484 B | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | devjiwonchoi/next.js refactor/cna | Change |
| - | - | - | - |
| index.html gzip | 530 B | 528 B | N/A |
| link.html gzip | 541 B | 540 B | N/A |
| withRouter.html gzip | 525 B | 522 B | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Edge SSR bundle Size</strong></summary>

|  | vercel/next.js canary  | devjiwonchoi/next.js refactor/cna | Change |
| - | - | - | - |
| edge-ssr.js gzip | 95.3 kB | 95.3 kB | N/A |
| page.js gzip | 3.04 kB | 3.04 kB | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Middleware size</strong></summary>

|  | vercel/next.js canary  | devjiwonchoi/next.js refactor/cna | Change |
| - | - | - | - |
| middleware-b..fest.js gzip | 627 B | 625 B | N/A |
| middleware-r..fest.js gzip | 151 B | 151 B | ✓ |
| middleware.js gzip | 25.5 kB | 25.5 kB | N/A |
| edge-runtime..pack.js gzip | 839 B | 839 B | ✓ |
| Overall change | 990 B | 990 B | ✓ |

</details>

<details>
<summary><strong>Next Runtimes</strong></summary>

|  | vercel/next.js canary  | devjiwonchoi/next.js refactor/cna | Change |
| - | - | - | - |
| app-page-exp...dev.js gzip | 170 kB | 170 kB | ✓ |
| app-page-exp..prod.js gzip | 97 kB | 97 kB | ✓ |
| app-page-tur..prod.js gzip | 98.8 kB | 98.8 kB | ✓ |
| app-page-tur..prod.js gzip | 93 kB | 93 kB | ✓ |
| app-page.run...dev.js gzip | 144 kB | 144 kB | ✓ |
| app-page.run..prod.js gzip | 91.5 kB | 91.5 kB | ✓ |
| app-route-ex...dev.js gzip | 21.4 kB | 21.4 kB | ✓ |
| app-route-ex..prod.js gzip | 15.1 kB | 15.1 kB | ✓ |
| app-route-tu..prod.js gzip | 15.1 kB | 15.1 kB | ✓ |
| app-route-tu..prod.js gzip | 14.8 kB | 14.8 kB | ✓ |
| app-route.ru...dev.js gzip | 21 kB | 21 kB | ✓ |
| app-route.ru..prod.js gzip | 14.8 kB | 14.8 kB | ✓ |
| pages-api-tu..prod.js gzip | 9.55 kB | 9.55 kB | ✓ |
| pages-api.ru...dev.js gzip | 9.82 kB | 9.82 kB | ✓ |
| pages-api.ru..prod.js gzip | 9.55 kB | 9.55 kB | ✓ |
| pages-turbo...prod.js gzip | 22.5 kB | 22.5 kB | ✓ |
| pages.runtim...dev.js gzip | 23.1 kB | 23.1 kB | ✓ |
| pages.runtim..prod.js gzip | 22.4 kB | 22.4 kB | ✓ |
| server.runti..prod.js gzip | 51 kB | 51 kB | ✓ |
| Overall change | 944 kB | 944 kB | ✓ |

</details>

<details>
<summary><strong>build cache</strong></summary>

|  | vercel/next.js canary  | devjiwonchoi/next.js refactor/cna | Change |
| - | - | - | - |
| 0.pack gzip | 1.58 MB | 1.58 MB | N/A |
| index.pack gzip | 106 kB | 105 kB | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Diff details</strong></summary>

<details>
<summary>Diff for <strong>middleware.js</strong></summary>

Diff too large to display
</details>

</details>

</details>
Commit: 7129c015c5d834b5db9ae960695c0576414913fd

**Comment by devjiwonchoi** [comment]
I'm not resolving package.json conflict because it will continuously conflict with the canary version update.

**Comment by devjiwonchoi** [comment]
will add tests and remove change on pkgjson

**Comment by ijjk** [comment]
## Failing test suites 
Commit: 19a89711014dc3afb8d6f0880d43b5f996cab146

`pnpm test test/integration/prerender-export/test/index.test.js`

<details>
<summary>Expand output</summary>



  ● Test suite failed to run

    thrown: "Exceeded timeout of 60000 ms for a hook.
    Add a timeout value to this test to increase the timeout, if this is a long-running test. See https://jestjs.io/docs/api#testname-fn-timeout."

      219 |         })
      220 |
    > 221 |         afterAll(async () => {
          |         ^
      222 |           if (app) {
      223 |             await stopApp(app)
      224 |           }

      at afterAll (integration/prerender-export/test/index.test.js:221:9)
      at describe (integration/prerender-export/test/index.test.js:211:7)
      at integration/prerender-export/test/index.test.js:208:56
      at Object.describe (integration/prerender-export/test/index.test.js:207:1)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

File: packages/create-next-app/index.ts [hunk 2]
```diff
@@ -137,307 +111,358 @@ const program = new Commander.Command(packageJson.name)
...
+  await tryCreateNextApp({
+    appPath,
+    resolvedOpts,
+    conf,
+    preferences,
+    isDryRun,
+  })
+}
+
+async function tryCreateNextApp({
```
Comment (devjiwonchoi) [inline packages/create-next-app/index.ts:427]: Split create app tasks as a separate function and run them based on the conditions: `example`, `isCI`, and basic.

File: packages/create-next-app/index.ts [hunk 3]
```diff
@@ -137,307 +111,358 @@ const program = new Commander.Command(packageJson.name)
...
+    const dryRunData = {
+      appPath,
+      packageManager,
+      opts,
+      args,
+      preferences,
     }
+    console.log(JSON.stringify(dryRunData, null, 2))
+    return
   }
```
Comment (devjiwonchoi) [inline packages/create-next-app/index.ts:452]: if dry run, return a JSON data like:

```json
{
  "appPath": "/Users/devjiwonchoi/DevLife/Vercel/next.js/packages/create-next-app/my-app",
  "packageManager": "npm",
  "opts": {
    "app": true,
    "typescript": true,
    "javascript": false,
    "eslint": true,
    "tailwind": true,
    "srcDir": true,
    "importAlias": "@/*"
  },
  "args": [
    "--dry-run"
  ],
  "preferences": {
    "typescript": true,
    "eslint": true,
    "tailwind": true,
    "srcDir": true,
    "importAlias": "@/*"
  }
}
```

**Comment by samcx** [comment]
@devjiwonchoi Will circle back on this 🔜!

**Comment by devjiwonchoi** [comment]
on syncing with updates on canary, making this WIP

**Comment by devjiwonchoi** [comment]
@samcx Thank you for taking a look, synced with changes on canary!

**Comment by samcx** [comment]
@devjiwonchoi Looks like there are conflicts on this now, do you mind taking a look at those?

**Comment by devjiwonchoi** [comment]
@samcx Resolved it, thank you!!

**Comment by samcx** [comment]
@devjiwonchoi I think I prefer the bold around **App Router**, **TypeScript**, **ESLint**, etc.

I also wonder if it's worth doing `'next/dist/compiled/commander' like for the Next.js CLI, in a separate PR <picture data-single-emoji=":frog-eyes:" title=":frog-eyes:">:frog-eyes:</picture>

**Comment by devjiwonchoi** [comment]
@samcx 

1. Bold over the names, got it!
2. Yeah, reducing it seems great, will open one on the following.
3. The Turbopack color with red/blue was intended as those are the primary colors:

> Tried to make these colors look catchy for the words 😄

 image

**Comment by leerob** [comment]
We might consider removing (recommended) from App Router, not that we _don't_ recommend it, just to not make it feel as "forced" or like the Pages Router is somehow deprecated (it's not).

**Comment by devjiwonchoi** [comment]
Removed (recommended) for App Router and reverted saving preference for it!

File: docs/01-getting-started/01-installation.mdx [hunk 4]
```diff
@@ -25,13 +25,13 @@ On installation, you'll see the following prompts:
 
 ```txt filename="Terminal"
 What is your project named? my-app
+Would you like to use App Router? No / Yes
```
Comment (devjiwonchoi) [inline docs/01-getting-started/01-installation.mdx:28]:
```suggestion
Would you like to use the App Router? No / Yes
```

File: docs/02-app/02-api-reference/06-create-next-app.mdx [hunk 5]
```diff
@@ -32,14 +32,15 @@ bun create next-app
...
-What is your project named?  my-app
-Would you like to use TypeScript?  No / Yes
-Would you like to use ESLint?  No / Yes
-Would you like to use Tailwind CSS?  No / Yes
-Would you like your code inside a `src/` directory?  No / Yes
-Would you like to use App Router? (recommended)  No / Yes
-Would you like to use Turbopack for `next dev`?  No / Yes
-Would you like to customize the import alias (`@/*` by default)?  No / Yes
+What is your project named? my-app
+Would you like to use App Router? No / Yes
```
Comment (devjiwonchoi) [inline docs/02-app/02-api-reference/06-create-next-app.mdx:36]:
```suggestion
Would you like to use the App Router? No / Yes
```

**Comment by devjiwonchoi** [comment]
cc @samcx 
Will stack the PRs on the following.

**Pull request closed without merging**